### PR TITLE
Kiln finalized tag

### PR DESCRIPTION
**Pull request by Hrugved**
## Changes:
-  Add finalized and safe block hash
- Add metadata DB to BlockTree
- Remove blockConfimationManager from ForkchoiceUpdatedV1Handler

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc/Modules/BlockFinderExtensions.cs [hunk 1]
```diff
@@ -76,9 +76,23 @@ public static SearchResult<Block> SearchForBlock(this IBlockFinder blockFinder,
                 block = blockFinder.FindBlock(blockParameter);
             }
 
-            return block == null && !allowNulls
-                ? new SearchResult<Block>($"Block {blockParameter.BlockHash?.ToString() ?? blockParameter.BlockNumber?.ToString() ?? blockParameter.Type.ToString()} could not be found", ErrorCodes.ResourceNotFound)
-                : new SearchResult<Block>(block);
+            if (block != null)
+            {
+                if (!block.IsPostMerge && (blockParameter.Equals(BlockParameter.Finalized) ||
```
Comment (Hrugved) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BlockFinderExtensions.cs:81]: how to determine if it's pre-merge? is this okay?

Comment (MarekM25) [reply src/Nethermind/Nethermind.JsonRpc/Modules/BlockFinderExtensions.cs:81]: Finalized is zero hash before the merge

File: src/Nethermind/Nethermind.JsonRpc/Modules/BlockFinderExtensions.cs [hunk 2]
```diff
@@ -76,9 +76,23 @@ public static SearchResult<Block> SearchForBlock(this IBlockFinder blockFinder,
...
+                                           blockParameter.Equals(BlockParameter.Safe)))
+                {
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.JsonRpc/Modules/BlockFinderExtensions.cs:83]: SafeHash can be Keccak.Zero too

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs [hunk 3]
```diff
@@ -491,6 +494,64 @@ public async Task forkchoiceUpdatedV1_should_work_with_zero_keccak_for_finalizat
...
+            blockForRpc.Should().NotBeNull();
+            actualFinalizedHash = blockForRpc.Hash;
+            actualFinalizedHash.Should().NotBeNull();
+            actualFinalizedHash.Should().Be(startingHead);
+            
+            AssertExecutionStatusChanged(rpc, newHeadHash!, startingHead, startingHead);
+        }
+        
+        [Test]
+        public async Task forkchoiceUpdatedV1_should_update_safe_block_hash()
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs:527]: we need to write tests with BlockFinder or EthRpcModuleTests (even better)

File: src/Nethermind/Nethermind.JsonRpc/Modules/BlockFinderExtensions.cs (hunk 1 above)
Comment (Hrugved) [reply src/Nethermind/Nethermind.JsonRpc/Modules/BlockFinderExtensions.cs:81]: https://github.com/ethereum/execution-apis/pull/200/files#:~:text=Before%20the%20merge,number%20or%20tag%22%2C

File: src/Nethermind/Nethermind.JsonRpc/Modules/BlockFinderExtensions.cs (hunk 2 above)
Comment (Hrugved) [reply src/Nethermind/Nethermind.JsonRpc/Modules/BlockFinderExtensions.cs:83]: https://github.com/ethereum/execution-apis/pull/200/files#:~:text=Before%20the%20merge,number%20or%20tag%22%2C

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs (hunk 3 above)
Comment (Hrugved) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V1.cs:527]: I already check in this test that rpc module returns the correct finalized/safe blocks. what else u suggest?

File: src/Nethermind/Nethermind.Blockchain/BlockTree.cs [hunk 4]
```diff
@@ -2094,5 +2095,17 @@ internal void ReleaseAcceptingNewBlocks()
...
+        public void ForkChoiceUpdated(Keccak? HeadBlockHash, Keccak? finalizedBlockHash, Keccak? safeBlockHash)
+        {
+            FinalizedHash = finalizedBlockHash;
+            SafeHash = safeBlockHash;
+            using (_metadataDb.StartBatch())
+            {
+                _metadataDb.Set(MetadataDbKeys.FinalizedBlockHash, FinalizedHash!.Bytes);  
+                _metadataDb.Set(MetadataDbKeys.SafeBlockHash, SafeHash!.Bytes);  
+            }
+            SetHeadBlock(HeadBlockHash);
```
Comment (Hrugved) [inline src/Nethermind/Nethermind.Blockchain/BlockTree.cs:2108]: should `FinalizedHash` and `SafeHash` be updated the same way as the `SetHeadBlock` function?

File: src/Nethermind/Nethermind.Merge.Plugin/Handlers/V1/ForkchoiceUpdatedV1Handler.cs [hunk 5]
```diff
@@ -166,6 +166,8 @@ public class ForkchoiceUpdatedV1Handler : IForkchoiceUpdatedV1Handler
                     _logger.Info(
                         $"Valid. ForkchoiceUpdated ignored - already in canonical chain. Request: {requestStr}");
                 }
+                
+                _blockTree.ForkChoiceUpdated(forkchoiceState.FinalizedBlockHash,forkchoiceState.SafeBlockHash);
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Merge.Plugin/Handlers/V1/ForkchoiceUpdatedV1Handler.cs:170]: remove this line, in this case we're ignoring fcU

**Review by MarekM25 — CHANGES_REQUESTED** [review]
Very good, pls check PoSSwitcher.ForkchoiceUpdated method. We're saving finalized hash. We should remove it from PoSSwitcher now :)

**Pull request merged**